File: ntl-systoolbox/src/modules/diagnostic.ts
import { exec } from "child_process";
import { promisify } from "util";
import mysql from "mysql2/promise";
import ora from "ora";
import chalk from "chalk";
import { Status, DiagnosticCheck, OutputResult, ExitCode } from "../types";
import { loadConfig } from "../utils/config";
import { getLogger } from "../utils/logger";
import { generateTimestamp, writeJsonOutput, printHumanOutput, determineExitCode, determineOverallStatus } from "../utils/output";

const execAsync = promisify(exec);

export async function runDiagnostic(action: string, inputs: any): Promise<void> {
  const checks: DiagnosticCheck[] = [];
  const anomalies: string[] = [];
  let spinner = ora("Démarrage du diagnostic...").start();

  try {
    switch (action) {
      case "ad_dns":
        await checkADDNS(checks, anomalies, spinner);
        break;
      case "mysql":
        await checkMySQL(checks, anomalies, spinner);
        break;
      case "windows":
        await checkWindowsServer(inputs.target, checks, anomalies, spinner);
        break;
      case "linux":
        await checkLinuxServer(inputs.target, checks, anomalies, spinner);
        break;
    }

    spinner.succeed(chalk.green("Diagnostic terminé"));

    const statuses = checks.map(c => c.status);
    const overallStatus = determineOverallStatus(statuses);
    const exitCode = determineExitCode(overallStatus);

    const result: OutputResult = {
      timestamp: generateTimestamp(),
      module: "diagnostic",
      status: overallStatus,
      exit_code: exitCode,
      summary: `Diagnostic ${action} completed with status ${overallStatus}`,
      details: { checks },
      anomalies: anomalies.length > 0 ? anomalies : undefined
    };

    writeJsonOutput(result);
    printHumanOutput(result);

    process.exitCode = exitCode;
  } catch (error: any) {
    spinner.fail(chalk.red("Échec du diagnostic"));
    throw error;
  }
}

async function checkADDNS(checks: DiagnosticCheck[], anomalies: string[], spinner: ora.Ora): Promise<void> {
  const config = loadConfig();
  
  for (const dc of config.ad.domain_controllers) {
    spinner.text = `Vérification du contrôleur de domaine: ${dc}`;
    
    try {
      const isWindows = process.platform === "win32";
      const pingCmd = isWindows ? `ping -n 2 ${dc}` : `ping -c 2 ${dc}`;
      await execAsync(pingCmd);
      
      checks.push({
        name: "AD Domain Controller",
        target: dc,
        status: Status.OK,
        message: `Contrôleur de domaine ${dc} accessible`
      });
      getLogger().info(`AD DC ${dc} is reachable`);
    } catch (error) {
      checks.push({
        name: "AD Domain Controller",
        target: dc,
        status: Status.CRIT,
        message: `Contrôleur de domaine ${dc} inaccessible`
      });
      anomalies.push(`DC ${dc} is unreachable`);
      getLogger().error(`AD DC ${dc} is unreachable`);
    }
  }

  for (const dns of config.ad.dns_servers) {
    spinner.text = `Vérification du serveur DNS: ${dns}`;
    
    try {
      const isWindows = process.platform === "win32";
      const dnsCmd = isWindows ? `nslookup ${config.ad.domain} ${dns}` : `dig @${dns} ${config.ad.domain}`;
      await execAsync(dnsCmd);
      
      checks.push({
        name: "DNS Server",
        target: dns,
        status: Status.OK,
        message: `Serveur DNS ${dns} fonctionnel`
      });
      getLogger().info(`DNS ${dns} is working`);
    } catch (error) {
      checks.push({
        name: "DNS Server",
        target: dns,
        status: Status.CRIT,
        message: `Serveur DNS ${dns} ne répond pas`
      });
      anomalies.push(`DNS ${dns} is not responding`);
      getLogger().error(`DNS ${dns} is not responding`);
    }
  }
}

async function checkMySQL(checks: DiagnosticCheck[], anomalies: string[], spinner: ora.Ora): Promise<void> {
  const config = loadConfig();
  spinner.text = `Connexion à MySQL: ${config.mysql.host}:${config.mysql.port}`;

  try {
    const connection = await mysql.createConnection({
      host: config.mysql.host,
      port: config.mysql.port,
      user: config.mysql.user,
      password: config.mysql.password
    });

    const [rows]: any = await connection.query("SELECT VERSION() as version");
    const version = rows[0].version;

    checks.push({
      name: "MySQL Connection",
      target: `${config.mysql.host}:${config.mysql.port}`,
      status: Status.OK,
      message: `MySQL accessible, version: ${version}`,
      metrics: { version }
    });

    const [databases]: any = await connection.query("SHOW DATABASES");
    const dbExists = databases.some((db: any) => db.Database === config.mysql.database);

    if (dbExists) {
      checks.push({
        name: "MySQL Database",
        target: config.mysql.database,
        status: Status.OK,
        message: `Base de données ${config.mysql.database} existe`
      });
    } else {
      checks.push({
        name: "MySQL Database",
        target: config.mysql.database,
        status: Status.WARN,
        message: `Base de données ${config.mysql.database} introuvable`
      });
      anomalies.push(`Database ${config.mysql.database} not found`);
    }

    await connection.end();
    getLogger().info("MySQL check completed successfully");
  } catch (error: any) {
    checks.push({
      name: "MySQL Connection",
      target: `${config.mysql.host}:${config.mysql.port}`,
      status: Status.CRIT,
      message: `Impossible de se connecter à MySQL: ${error.message}`
    });
    anomalies.push(`MySQL connection failed: ${error.message}`);
    getLogger().error("MySQL check failed", { error: error.message });
  }
}

async function checkWindowsServer(target: string, checks: DiagnosticCheck[], anomalies: string[], spinner: ora.Ora): Promise<void> {
  spinner.text = `Diagnostic Windows Server: ${target}`;

  try {
    if (target === "localhost" && process.platform === "win32") {
      const { stdout: osVersion } = await execAsync("ver");
      checks.push({
        name: "OS Version",
        target,
        status: Status.OK,
        message: osVersion.trim(),
        metrics: { os_version: osVersion.trim() }
      });

      const { stdout: uptime } = await execAsync("net statistics workstation | findstr \"since\"");
      checks.push({
        name: "Uptime",
        target,
        status: Status.OK,
        message: uptime.trim(),
        metrics: { uptime: uptime.trim() }
      });

      const { stdout: cpu } = await execAsync("wmic cpu get loadpercentage /value");
      const cpuMatch = cpu.match(/LoadPercentage=(\d+)/);
      const cpuUsage = cpuMatch ? parseInt(cpuMatch[1]) : 0;
      checks.push({
        name: "CPU Usage",
        target,
        status: cpuUsage > 80 ? Status.WARN : Status.OK,
        message: `Utilisation CPU: ${cpuUsage}%`,
        metrics: { cpu_usage_percent: cpuUsage }
      });

      const { stdout: mem } = await execAsync("wmic OS get FreePhysicalMemory,TotalVisibleMemorySize /value");
      const totalMatch = mem.match(/TotalVisibleMemorySize=(\d+)/);
      const freeMatch = mem.match(/FreePhysicalMemory=(\d+)/);
      if (totalMatch && freeMatch) {
        const totalMem = parseInt(totalMatch[1]) / 1024;
        const freeMem = parseInt(freeMatch[1]) / 1024;
        const usedPercent = ((totalMem - freeMem) / totalMem) * 100;
        checks.push({
          name: "Memory Usage",
          target,
          status: usedPercent > 90 ? Status.WARN : Status.OK,
          message: `Mémoire utilisée: ${usedPercent.toFixed(1)}%`,
          metrics: { memory_usage_percent: usedPercent, total_mb: totalMem, free_mb: freeMem }
        });
      }

      const { stdout: disk } = await execAsync("wmic logicaldisk get size,freespace,caption");
      checks.push({
        name: "Disk Space",
        target,
        status: Status.OK,
        message: "Informations disque récupérées",
        metrics: { disk_info: disk.trim() }
      });

      getLogger().info(`Windows Server diagnostic completed for ${target}`);
    } else {
      checks.push({
        name: "Windows Server Check",
        target,
        status: Status.WARN,
        message: "Diagnostic distant non implémenté (nécessite WinRM/PowerShell Remoting)"
      });
      anomalies.push("Remote Windows diagnostics require WinRM configuration");
    }
  } catch (error: any) {
    checks.push({
      name: "Windows Server Check",
      target,
      status: Status.CRIT,
      message: `Erreur: ${error.message}`
    });
    anomalies.push(`Windows check failed: ${error.message}`);
    getLogger().error("Windows check failed", { error: error.message });
  }
}

async function checkLinuxServer(target: string, checks: DiagnosticCheck[], anomalies: string[], spinner: ora.Ora): Promise<void> {
  spinner.text = `Diagnostic Linux: ${target}`;

  try {
    if (target === "localhost" && process.platform !== "win32") {
      const { stdout: osVersion } = await execAsync("cat /etc/os-release | grep PRETTY_NAME");
      checks.push({
        name: "OS Version",
        target,
        status: Status.OK,
        message: osVersion.trim(),
        metrics: { os_version: osVersion.trim() }
      });

      const { stdout: uptime } = await execAsync("uptime -p");
      checks.push({
        name: "Uptime",
        target,
        status: Status.OK,
        message: uptime.trim(),
        metrics: { uptime: uptime.trim() }
      });

      const { stdout: cpu } = await execAsync("top -bn1 | grep \"Cpu(s)\" | awk '{print $2}'");
      const cpuUsage = parseFloat(cpu.trim());
      checks.push({
        name: "CPU Usage",
        target,
        status: cpuUsage > 80 ? Status.WARN : Status.OK,
        message: `Utilisation CPU: ${cpuUsage}%`,
        metrics: { cpu_usage_percent: cpuUsage }
      });

      const { stdout: mem } = await execAsync("free -m | grep Mem");
      const memParts = mem.trim().split(/\s+/);
      const totalMem = parseInt(memParts[1]);
      const usedMem = parseInt(memParts[2]);
      const usedPercent = (usedMem / totalMem) * 100;
      checks.push({
        name: "Memory Usage",
        target,
        status: usedPercent > 90 ? Status.WARN : Status.OK,
        message: `Mémoire utilisée: ${usedPercent.toFixed(1)}%`,
        metrics: { memory_usage_percent: usedPercent, total_mb: totalMem, used_mb: usedMem }
      });

      const { stdout: disk } = await execAsync("df -h /");
      checks.push({
        name: "Disk Space",
        target,
        status: Status.OK,
        message: "Informations disque récupérées",
        metrics: { disk_info: disk.trim() }
      });

      getLogger().info(`Linux diagnostic completed for ${target}`);
    } else {
      checks.push({
        name: "Linux Check",
        target,
        status: Status.WARN,
        message: "Diagnostic distant non implémenté (nécessite SSH)"
      });
      anomalies.push("Remote Linux diagnostics require SSH configuration");
    }
  } catch (error: any) {
    checks.push({
      name: "Linux Check",
      target,
      status: Status.CRIT,
      message: `Erreur: ${error.message}`
    });
    anomalies.push(`Linux check failed: ${error.message}`);
    getLogger().error("Linux check failed", { error: error.message });
  }
}

File: src/ntlsystoolbox/modules/diagnostic.py
"""Module Diagnostic - Vérifications AD/DNS/MySQL/OS"""

import subprocess
import platform
from typing import Dict, Any
import pymysql

from ntlsystoolbox.utils.output import (
    display_result, save_json_report, aggregate_status, status_to_exitcode
)


def check_ad_dns(config: Dict[str, Any]) -> int:
    """Vérifie la disponibilité des contrôleurs AD et serveurs DNS"""
    anomalies = []
    details = {}
    statuses = []
    
    domain_controllers = config.get("ad", {}).get("domain_controllers", [])
    dns_servers = config.get("dns", {}).get("servers", [])
    
    for dc in domain_controllers:
        if not dc:
            continue
        try:
            result = subprocess.run(
                ["ping", "-n" if platform.system() == "Windows" else "-c", "1", dc],
                capture_output=True,
                timeout=5
            )
            if result.returncode == 0:
                details[f"AD_{dc}"] = "OK"
                statuses.append("OK")
            else:
                details[f"AD_{dc}"] = "UNREACHABLE"
                anomalies.append(f"Contrôleur AD {dc} injoignable")
                statuses.append("CRIT")
        except Exception as e:
            details[f"AD_{dc}"] = f"ERROR: {e}"
            anomalies.append(f"Erreur AD {dc}: {e}")
            statuses.append("CRIT")
    
    for dns in dns_servers:
        if not dns:
            continue
        try:
            result = subprocess.run(
                ["nslookup", "google.com", dns] if platform.system() == "Windows" 
                else ["dig", "@" + dns, "google.com"],
                capture_output=True,
                timeout=5
            )
            if result.returncode == 0:
                details[f"DNS_{dns}"] = "OK"
                statuses.append("OK")
            else:
                details[f"DNS_{dns}"] = "FAIL"
                anomalies.append(f"Serveur DNS {dns} ne répond pas")
                statuses.append("WARN")
        except Exception as e:
            details[f"DNS_{dns}"] = f"ERROR: {e}"
            anomalies.append(f"Erreur DNS {dns}: {e}")
            statuses.append("WARN")
    
    status = aggregate_status(statuses) if statuses else "OK"
    summary = f"Vérifications AD/DNS: {len([s for s in statuses if s == 'OK'])}/{len(statuses)} OK"
    
    display_result("diagnostic_ad_dns", status, summary, details, anomalies)
    save_json_report("diagnostic_ad_dns", status, summary, details, anomalies, 
                     config.get("reports", {}).get("path", "./reports"))
    
    return status_to_exitcode(status)


def check_mysql(config: Dict[str, Any]) -> int:
    """Teste la connexion et l'état de la base MySQL"""
    anomalies = []
    details = {}
    
    mysql_cfg = config.get("mysql", {})
    
    try:
        conn = pymysql.connect(
            host=mysql_cfg.get("host", "localhost"),
            port=mysql_cfg.get("port", 3306),
            user=mysql_cfg.get("user", "root"),
            password=mysql_cfg.get("password", ""),
            database=mysql_cfg.get("database", "wms"),
            connect_timeout=10
        )
        
        cursor = conn.cursor()
        
        cursor.execute("SELECT VERSION()")
        version = cursor.fetchone()[0]
        details["version"] = version
        
        cursor.execute("SHOW DATABASES")
        databases = [row[0] for row in cursor.fetchall()]
        details["databases_count"] = len(databases)
        details["databases"] = databases
        
        cursor.close()
        conn.close()
        
        status = "OK"
        summary = f"MySQL {mysql_cfg.get('host')} accessible, version {version}"
        
    except Exception as e:
        status = "CRIT"
        summary = f"Erreur connexion MySQL: {e}"
        anomalies.append(str(e))
        details["error"] = str(e)
    
    display_result("diagnostic_mysql", status, summary, details, anomalies)
    save_json_report("diagnostic_mysql", status, summary, details, anomalies,
                     config.get("reports", {}).get("path", "./reports"))
    
    return status_to_exitcode(status)


def check_windows_server(hostname: str, config: Dict[str, Any]) -> int:
    """Diagnostic d'un serveur Windows"""
    anomalies = []
    details = {"hostname": hostname}
    statuses = []
    
    try:
        result = subprocess.run(["systeminfo"], capture_output=True, text=True, timeout=30)
        output = result.stdout
        
        for line in output.split("\n"):
            if "OS Name" in line:
                details["os_name"] = line.split(":", 1)[1].strip()
            elif "OS Version" in line:
                details["os_version"] = line.split(":", 1)[1].strip()
            elif "System Up Time" in line:
                details["uptime"] = line.split(":", 1)[1].strip()
        
        statuses.append("OK")
        
    except Exception as e:
        anomalies.append(f"Erreur systeminfo: {e}")
        details["systeminfo_error"] = str(e)
        statuses.append("WARN")
    
    try:
        result = subprocess.run(
            ["wmic", "cpu", "get", "loadpercentage"],
            capture_output=True, text=True, timeout=10
        )
        cpu_lines = [l.strip() for l in result.stdout.split("\n") if l.strip().isdigit()]
        if cpu_lines:
            cpu_usage = int(cpu_lines[0])
            details["cpu_usage_percent"] = cpu_usage
            if cpu_usage > 90:
                anomalies.append(f"CPU élevé: {cpu_usage}%")
                statuses.append("WARN")
            else:
                statuses.append("OK")
    except Exception as e:
        details["cpu_error"] = str(e)
        statuses.append("WARN")
    
    status = aggregate_status(statuses)
    summary = f"Diagnostic Windows {hostname}: {details.get('os_name', 'N/A')}"
    
    display_result("diagnostic_windows", status, summary, details, anomalies)
    save_json_report("diagnostic_windows", status, summary, details, anomalies,
                     config.get("reports", {}).get("path", "./reports"))
    
    return status_to_exitcode(status)


def check_linux_server(hostname: str, config: Dict[str, Any]) -> int:
    """Diagnostic d'un serveur Linux"""
    anomalies = []
    details = {"hostname": hostname}
    statuses = []
    
    try:
        with open("/etc/os-release", "r") as f:
            for line in f:
                if line.startswith("PRETTY_NAME"):
                    details["os_name"] = line.split("=")[1].strip().strip('"')
        statuses.append("OK")
    except Exception as e:
        details["os_error"] = str(e)
        statuses.append("WARN")
    
    try:
        result = subprocess.run(["uptime"], capture_output=True, text=True, timeout=5)
        details["uptime"] = result.stdout.strip()
        statuses.append("OK")
    except Exception as e:
        details["uptime_error"] = str(e)
        statuses.append("WARN")
    
    try:
        result = subprocess.run(
            ["top", "-bn1"], capture_output=True, text=True, timeout=5
        )
        for line in result.stdout.split("\n"):
            if "Cpu(s)" in line:
                details["cpu_info"] = line.strip()
                break
        statuses.append("OK")
    except Exception as e:
        details["cpu_error"] = str(e)
        statuses.append("WARN")
    
    status = aggregate_status(statuses)
    summary = f"Diagnostic Linux {hostname}: {details.get('os_name', 'N/A')}"
    
    display_result("diagnostic_linux", status, summary, details, anomalies)
    save_json_report("diagnostic_linux", status, summary, details, anomalies,
                     config.get("reports", {}).get("path", "./reports"))
    
    return status_to_exitcode(status)
